File: elixir/2-AddTwoNumbers/main.exs
defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  @spec add_two_numbers(l1 :: ListNode.t | nil, l2 :: ListNode.t | nil) :: ListNode.t | nil
  def add_two_numbers(l1, l2) do
    add(l1, l2, 0)
  end

  defp add(l1, l2, carry) when l1 != nil and l2 != nil do
    {digit, carry} = digit_carry(l1.val + l2.val + carry)
    %ListNode{val: digit, next: add(l1.next, l2.next, carry)}
  end

  defp add(l1, nil, 1) when l1 != nil do
    {digit, carry} = digit_carry(l1.val + 1)
    %ListNode{val: digit, next: add(l1.next, nil, carry)}
  end

  defp add(nil, l2, 1) when l2 != nil do
    {digit, carry} = digit_carry(l2.val + 1)
    %ListNode{val: digit, next: add(nil, l2.next, carry)}
  end

  defp add(nil, nil, 1) do
    %ListNode{val: 1}
  end

    defp add(l1, l2, 0) when l1 == nil or l2 == nil do
    if l1 == nil do
      l2
    else
      l1
    end
  end

  defp digit_carry(total) do
    digit = rem(total, 10)
    carry = div(total, 10)
    {digit, carry}
  end
end

defmodule Tests do
  defp list_to_string(list) when list == nil do
    ""
  end

  defp list_to_string(list) do
    "#{list.val} " <> list_to_string(list.next)
  end

  def example1 do
    list1 = %ListNode{val: 2, next: %ListNode{val: 4, next: %ListNode{val: 3}}}
    list2 = %ListNode{val: 5, next: %ListNode{val: 6, next: %ListNode{val: 4}}}
    IO.puts(list_to_string(Solution.add_two_numbers(list1, list2)))
  end

  def example2 do
    list1 = %ListNode{val: 0}
    list2 = %ListNode{val: 0}
    IO.puts(list_to_string(Solution.add_two_numbers(list1, list2)))
  end

  def example3 do
    list1 = %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9}}}}}}}
    list2 = %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9, next: %ListNode{val: 9}}}}
    IO.puts(list_to_string(Solution.add_two_numbers(list1, list2)))
  end
end

Tests.example1
Tests.example2
Tests.example3

File: rust/2-AddTwoNumbers/main.rs
 //Definition for singly-linked list.
 #[derive(PartialEq, Eq, Clone, Debug)]
 pub struct ListNode {
   pub val: i32,
   pub next: Option<Box<ListNode>>
}
 
impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    pub fn recurse(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>, carry: i32) -> Option<Box<ListNode>> {
        match (l1, l2) {
            (Some(x), Some(y)) => {
                let total = x.val + y.val + carry;
                return Some(Box::new(ListNode{val: total % 10, next: recurse(x.next, y.next, total / 10)}))
            }
            (Some(x), None) => {
                let total = x.val + carry;
                return Some(Box::new(ListNode{val: total % 10, next: recurse(x.next, None, total / 10)}))
            }
            (None, Some(y)) => {
                let total = y.val + carry;
                return Some(Box::new(ListNode{val: total % 10, next: recurse(None, y.next, total / 10)}))
            }

            (None, None) => {
                if carry == 0 {
                    None
                } else {
                    Some(Box::new(ListNode{val: carry, next: None}))
                }
            }
        }
    }

    recurse(l1, l2, 0)
}

fn main() {
    let list1 = Some(Box::new(ListNode{
        val: 1, 
        next: Some(Box::new(ListNode{
            val: 2, 
            next: Some(Box::new(ListNode{
                val: 3, 
                next: None
            }))
        }))
    }));

    let list2 = Some(Box::new(ListNode{
        val: 1, 
        next: Some(Box::new(ListNode{
            val: 2, 
            next: Some(Box::new(ListNode{
                val: 3, 
                next: None
            }))
        }))
    }));

    let result = add_two_numbers(list1, list2);
}